### DISCUSSION TITLE
How to create a ValueTree object?

### DISCUSSION THREAD

**Discussion by dobo365** [RomanKubiak/ctrlr#405] [Q&A]
Using 5.3.201...
This simple statement should work but I get Stack overflow... Any idea?

```
vt = ValueTree("data")

ERROR: [string "ValueTree"]:3: stack overflow
```

**Comment by damiensellier** [comment]
I tried a button triggering a LUA with 

vt = ValueTree("data")

When the button value change I don't get the ERROR
Can you please share a sample panel that gives this error?

EDIT: From Juce documentation about the [ValueTree Class ](https://docs.juce.com/master/classValueTree.html#ad5671081941b3db9db42bf7fcbd2629c) and from the CtrlrValueTreeEditor file, it seems that a ValueTree needs a unique name on construction and a content probably as a LUA table.

ValueTree("data") is probably not enough and needs somerhing like saveState, readStade.
So I would go like with the readState : 

vt =  ValueTree("ValueTreeID")
vt:setProperty("PropertyName", propertyValue, nil)

**Reply by damiensellier** [reply]
I got it working on all Ctrlr Versions I have (5.3.198 and 5.6.23), it's exactly like the saveState Function that is meant to create a ValueTree with a predefined nameID : 

[ValueTree Write Read_1_0_0_2022-01-09_19-04.panel.zip](https://github.com/RomanKubiak/ctrlr/files/7836244/ValueTree.Write.Read_1_0_0_2022-01-09_19-04.panel.zip)

The 1st button construct the tree and set a value for a certain property

```
vt = ValueTree("ValueTreeID")
vt:setProperty("PropertyName", 1, nil)
```

The 2nd button outputs the value from the tree in the console

```
if vt:hasProperty("PropertyName") == true then 
	console("you are here")
	vtPropertyValue = vt:getProperty("PropertyName")
        console("This is the PropertynameValue : "..(tostring(vtPropertyValue)))
end -- if

```
And the XML created will be like : 
`<ValueTreeID PropertyName="propertyValue"/>`

**Reply by Godlike-Productions** [reply]
In your example here is stateData a pre-declared valueTree object?  In your example, I don't see where you declare it

ie I'd expect
```
stateData = valueTree("stateData")
```

**Reply by Godlike-Productions** [reply]
I've just built a method to save and restore all of the MIDI parameters in the MIDI settings box.  To save everyone a bunch of time (and so I can find it later) here's the methods.

Note: I've included the last path that I saved files to here as well.  Just delete this line if you don't need it.

Save to Value Tree

```
--
-- Called when data needs saving
--
saveStateData = function(--[[ ValueTree --]] stateData)

saveAllBanks()

-- Update ValueTree
savePath = tostring(panel:getLabelComponent("filePathLabel"):getProperty("uiLabelText"))
--savePath = "Testing"
sessionMidiInDevice = panel:getProperty("panelMidiInputDevice")
sessionMidiOutDevice = panel:getProperty("panelMidiOutputDevice")
sessionMidiControlDevice = panel:getProperty("panelMidiControllerDevice")
sessionMidiInDeviceChannel = panel:getProperty("panelMidiInputChannelDevice")
sessionMidiOutDeviceChannel = panel:getProperty("panelMidiOutputChannelDevice")
sessionMidiControlDeviceChannel = panel:getProperty("panelMidiControllerChannelDevice")
sessionMidiInputFromHost = panel:getProperty("panelMidiInputFromHost")
sessionMidiInputFromHostChannel = panel:getProperty("panelMidiInputChannelHost")
sessionMidiOutputToHost = panel:getProperty("panelMidiOutputToHost")
sessionMidiOutputToHostChannel = panel:getProperty("panelMidiOutputChannelHost")
sessionMidiThruH2H = panel:getProperty("panelMidiThruH2H")
sessionMidiThruH2HChannel = panel:getProperty("panelMidiThruH2HChannelize")
sessionMidiThruH2D = panel:getProperty("panelMidiThruH2D")
sessionMidiThruH2DChannel = panel:getProperty("panelMidiThruH2DChannelize")
sessionMidiThruD2D = panel:getProperty("panelMidiThruD2D")
sessionMidiThruD2DChannel = panel:getProperty("panelMidiThruD2DChannelize")
sessionMidiThruD2H = panel:getProperty("panelMidiThruD2H")
sessionMidiThruD2HChannel = panel:getProperty("panelMidiThruD2HChannelize")

--vtTest = ValueTree("stateData2")

stateData:setProperty("lastPath", savePath, nil)
--stateData:addChild(vtTest, -1, nil)
stateData:setProperty("lastPath", savePath, nil)
stateData:setProperty("midiInDevice", sessionMidiInDevice, nil)
stateData:setProperty("midiOutDevice", sessionMidiOutDevice, nil)
stateData:setProperty("midiControlDevice", sessionMidiControlDevice, nil)
stateData:setProperty("midiInDeviceChannel", sessionMidiInDeviceChannel, nil)
stateData:setProperty("midiOutDeviceChannel", sessionMidiOutDeviceChannel, nil)
stateData:setProperty("midiControlDeviceChannel", sessionMidiControlDeviceChannel, nil)
stateData:setProperty("midiInputFromHost", sessionMidiInputFromHost, nil)
stateData:setProperty("midiInputFromHostChannel", sessionMidiInputFromHostChannel, nil)
stateData:setProperty("midiOutputToHost", sessionMidiOutputToHost, nil)
stateData:setProperty("midiOutputToHostChannel", sessionMidiOutputToHostChannel, nil)
stateData:setProperty("midiThruH2H", sessionmidiThruH2H, nil)
stateData:setProperty("midiThruH2HChannel", sessionmidiThruH2HChannel, nil)
stateData:setProperty("midiThruH2D", sessionmidiThruH2D, nil)
stateData:setProperty("midiThruH2DChannel", sessionmidiThruH2DChannel, nil)
stateData:setProperty("midiThruD2D", sessionmidiThruD2D, nil)
stateData:setProperty("midiThruD2DChannel", sessionmidiThruD2DChannel, nil)
stateData:setProperty("midiThruD2H", sessionmidiThruD2H, nil)
stateData:setProperty("midiThruD2HChannel", sessionmidiThruD2HChannel, nil)


end
```

And to reload

```
--
-- Called when data is restored
--
loadStateData = function(--[[ ValueTree --]] stateData)

-- Read From ValueTree

if stateData:hasProperty("lastPath") then
    savePath = stateData:getProperty("lastPath")
    panel:getLabelComponent("filePathLabel"):setText(savePath)
else
    savePath = L(File.getSpecialLocation(File.userDocumentsDirectory):getFullPathName())
             if (savePath:sub(-1) ~= '/' and savePath:sub(-1) ~= '\\') then
                savePath = savePath..'/'
            end --if
    panel:getLabelComponent("filePathLabel"):setText (savePath)
end --if

if stateData:hasProperty("midiInDevice") then
    sessionMidiInDevice = stateData:getProperty("midiInDevice")
    panel:setProperty("panelMidiInputDevice",sessionMidiInDevice)
end --if

if stateData:hasProperty("midiOutDevice") then
    sessionMidiOutDevice = stateData:getProperty("midiOutDevice")
    panel:setProperty("panelMidiOutputDevice",sessionMidiOutDevice)
end --if

if stateData:hasProperty("midiControlDevice") then
    sessionMidiControlDevice = stateData:getProperty("midiControlDevice")
    panel:setProperty("panelMidiControllerDevice",sessionMidiControlDevice)
end --if


if stateData:hasProperty("midiInDeviceChannel") then
    sessionMidiInDeviceChannel = stateData:getProperty("midiInDeviceChannel")
    panel:setProperty("panelMidiInputChannelDevice",sessionMidiInDeviceChannel)
end --if

if stateData:hasProperty("midiOutDeviceChannel") then
    sessionMidiOutDeviceChannel = stateData:getProperty("midiOutDeviceChannel")
    panel:setProperty("panelMidiOutputChannelDevice",sessionMidiOutDeviceChannel)
end --if

if stateData:hasProperty("midiControlDeviceChannel") then
    sessionMidiControlDeviceChannel = stateData:getProperty("midiControlDeviceChannel")
    panel:setProperty("panelMidiControllerChannelDevice",sessionMidiControlDeviceChannel)
end --if

if stateData:hasProperty("midiInputFromHost") then
    sessionMidiInputFromHost = stateData:getProperty("midiInputFromHost")
    panel:setProperty("panelMidiInputFromHost",sessionMidiInputFromHost)
end --if

if stateData:hasProperty("midiInputFromHostChannel") then
    sessionMidiInputFromHostChannel = stateData:getProperty("midiInputFromHostChannel")
    panel:setProperty("panelMidiInputChannelHost",sessionMidiInputFromHostChannel)
end --if

if stateData:hasProperty("midiOutputToHost") then
    sessionMidiOutputToHost = stateData:getProperty("midiOutputToHost")
    panel:setProperty("panelMidiOutputToHost",sessionMidiOutputToHost)
end --if

if stateData:hasProperty("midiOutputToHostChannel") then
    sessionMidiOutputToHostChannel = stateData:getProperty("midiOutputToHostChannel")
    panel:setProperty("panelMidiOutputChannelHost",sessionMidiOutputToHostChannel)
end --if

if stateData:hasProperty("midiThruH2H") then
    sessionMidiThruH2H = stateData:getProperty("midiThruH2H")
    panel:setProperty("panelMidiThruH2H",sessionMidiThruH2H)
end --if

if stateData:hasProperty("midiThruH2HChannel") then
    sessionMidiThruH2HChannel = stateData:getProperty("midiThruH2HChannel")
    panel:setProperty("panelMidiThruH2HChannelize",sessionMidiThruH2HChannel)
end --if

if stateData:hasProperty("midiThruH2D") then
    sessionMidiThruH2D = stateData:getProperty("midiThruH2D")
    panel:setProperty("panelMidiThruH2D",sessionMidiThruH2D)
end --if

if stateData:hasProperty("midiThruH2DChannel") then
    sessionMidiThruH2DChannel = stateData:getProperty("midiThruH2DChannel")
    panel:setProperty("panelMidiThruH2DChannelize",sessionMidiThruH2DChannel)
end --if

if stateData:hasProperty("midiThruD2D") then
    sessionMidiThruD2D = stateData:getProperty("midiThruD2D")
    panel:setProperty("panelMidiThruD2D",sessionMidiThruD2D)
end --if

if stateData:hasProperty("midiThruD2DChannel") then
    sessionMidiThruD2DChannel = stateData:getProperty("midiThruD2DChannel")
    panel:setProperty("panelMidiThruD2DChannelize",sessionMidiThruD2DChannel)
end --if

if stateData:hasProperty("midiThruD2H") then
    sessionMidiThruD2H = stateData:getProperty("midiThruD2H")
    panel:setProperty("panelMidiThruD2H",sessionMidiThruD2H)
end --if

if stateData:hasProperty("midiThruD2HChannel") then
    sessionMidiThruD2HChannel = stateData:getProperty("midiThruD2HChannel")
    panel:setProperty("panelMidiThruD2HChannelize",sessionMidiThruD2HChannel)
end --if


loadAllBanks()

end
```

**Reply by damiensellier** [reply]
> In your example here is stateData a pre-declared valueTree object? In your example, I don't see where you declare it
> 
> ie I'd expect
> 
> ```
> stateData = valueTree("stateData")
> ```

stateData is a function within Ctrlr that covers the JUCE valuetree class with the destination valuetree pre-defined by the Ctrlr function tself as an XML string called "panelCustomData" at the end of Ctrlr.settings in : username/Library/Ctrlr/Ctrlr.settings
So stateData is a preset valuetree with the property predefined (panelCustomData) and the file destination Ctrlr.settings.
And everytime the Ctrlr app launches there's a function readStateData that tries to grab the values from this XML string.
That's smart!

**Reply by Godlike-Productions** [reply]
OK, cool. I'll change my examples to load and save to that.  Loading from my examples causes a crash.

**Reply by damiensellier** [reply]
> OK, cool. I'll change my examples to load and save to that. Loading from my examples causes a crash.

Your function is really nice and should be very handy for exemple if you want to save a configuration as a file. SaveState is perfect for automatic state restoration when the panel is shutdown/loaded.

**Reply by Godlike-Productions** [reply]
Is there a better way to do what I've done automatically, using some sort of saveState command.  I only really want to save the file path and the MIDI settings.  Saving the modulator state really makes no sense with my panel at all (as I load them from the synth's front panel)

**Reply by damiensellier** [reply]
> Is there a better way to do what I've done automatically, using some sort of saveState command. I only really want to save the file path and the MIDI settings. Saving the modulator state really makes no sense with my panel at all (as I load them from the synth's front panel)

Sure it's super easy. Just follow @dobo365 Dominique's explanation from the Ctrlr.org forum here : https://ctrlr.org/forums/topic/global-persistent-variables/#post-118293
This way you can store all the paths to stateData
As mentioned before it's only working when the panels are shut down as plugin in the host (the stateData will be stored in the host session file, ex for cubase : SongName.cpr) OR within the ctrlr.app if you close the app, not "close"  the panel, or you'll erase those state values. (saved in Ctrlr.settings as custom nodes in panelCustomData)

**Reply by Godlike-Productions** [reply]
I'm having no luck with this.

This is my method -- Please see post above for code that works.  This stuff doesn't, but kept here for thread integrity

```
--
-- Called when data needs saving
--
saveStateData = function(--[[ ValueTree --]] stateData)

-- Update ValueTree
savePath = L(panel:getLabelComponent("filePathLabel"):getPropertyString("uiLabelText"))
sessionMidiInDevice = panel:getPropertyString("panelMidiInputDevice")
sessionMidiOutDevice = panel:getPropertyString("panelMidiOutputDevice")
sessionMidiControlDevice = panel:getPropertyString("panelMidiControllerDevice")
sessionMidiInDeviceChannel = panel:getPropertyString("panelMidiInputChannelDevice")
sessionMidiOutDeviceChannel = panel:getPropertyString("panelMidiOutputChannelDevice")
sessionMidiControlDeviceChannel = panel:getPropertyString("panelMidiControllerChannelDevice")
sessionMidiInputFromHost = panel:getPropertyString("panelMidiInputFromHost")
sessionMidiInputFromHostChannel = panel:getPropertyString("panelMidiInputChannelHost")
sessionMidiOutputToHost = panel:getPropertyString("panelMidiOutputToHost")
sessionMidiOutputToHostChannel = panel:getPropertyString("panelMidiOutputChannelHost")
sessionMidiThruH2H = panel:getPropertyString("panelMidiThruH2H")
sessionMidiThruH2HChannel = panel:getPropertyString("panelMidiThruH2HChannelize")
sessionMidiThruH2D = panel:getPropertyString("panelMidiThruH2D")
sessionMidiThruH2DChannel = panel:getPropertyString("panelMidiThruH2DChannelize")
sessionMidiThruD2D = panel:getPropertyString("panelMidiThruD2D")
sessionMidiThruD2DChannel = panel:getPropertyString("panelMidiThruD2DChannelize")
sessionMidiThruD2H = panel:getPropertyString("panelMidiThruD2H")
sessionMidiThruD2HChannel = panel:getPropertyString("panelMidiThruD2HChannelize")


-- stateData = ValueTree("stateData")
stateData:setProperty("lastPath", savePath, nil)
stateData:setProperty("midiInDevice", sessionMidiInDevice, nil)
stateData:setProperty("midiOutDevice", sessionMidiOutDevice, nil)
stateData:setProperty("midiControlDevice", sessionMidiControlDevice, nil)
stateData:setProperty("midiInDeviceChannel", sessionMidiInDeviceChannel, nil)
stateData:setProperty("midiOutDeviceChannel", sessionMidiOutDeviceChannel, nil)
stateData:setProperty("midiControlDeviceChannel", sessionMidiControlDeviceChannel, nil)
stateData:setProperty("midiInputFromHost", sessionMidiInputFromHost, nil)
stateData:setProperty("midiInputFromHostChannel", sessionMidiInputFromHostChannel, nil)
stateData:setProperty("midiOutputToHost", sessionMidiOutputToHost, nil)
stateData:setProperty("midiOutputToHostChannel", sessionMidiOutputToHostChannel, nil)
stateData:setProperty("midiThruH2H", sessionmidiThruH2H, nil)
stateData:setProperty("midiThruH2HChannel", sessionmidiThruH2HChannel, nil)
stateData:setProperty("midiThruH2D", sessionmidiThruH2D, nil)
stateData:setProperty("midiThruH2DChannel", sessionmidiThruH2DChannel, nil)
stateData:setProperty("midiThruD2D", sessionmidiThruD2D, nil)
stateData:setProperty("midiThruD2DChannel", sessionmidiThruD2DChannel, nil)
stateData:setProperty("midiThruD2H", sessionmidiThruD2H, nil)
stateData:setProperty("midiThruD2HChannel", sessionmidiThruD2HChannel, nil)


end
```

when I try to call it I get this in the console

```
>>> saveStateData()
ERROR: [string "saveStateData"]:29: attempt to index local 'stateData' (a nil value)

>>> saveStateData(stateData)
ERROR: [string "saveStateData"]:29: attempt to index local 'stateData' (a nil value)
```

Nothing seems to happen if I go to File -> Save Ctrlr State

**Reply by damiensellier** [reply]
@Godlike-Productions I would first try the saveState until the path instruction and check if you have an error and let's see exactly wher it breaks one by one.
Check also the format of the property, if it's a number or a string, an hexstring or a table or a MemoryBlock in general, then convert them to string for XML. and with readState put them back to their appropriate format with tonumber, tostring, tohexstring....

Also, check the midiDevice settings, they are not getPropertyString but just **getProperty**:
`panel:getProperty("panelMidiOutputDevice")`

Note : Don't forget getPropertyInt if the value property is an integer.

**Reply by dobo365** [reply]
Sorry, but why do you want to save and restore the Midi settings?
This is already taken care automatically by the panel...

You should use stateData when you want to keep other things (variables but also specific labels, tables...).
As explained before the stateData variable is predefined within Ctrlr and available in the methods using the templates to read and save data at panel opening/closing.
Pay attention that ReadStateData is performed 2 times. Once before panel loaded and once after.
Therefore, don't assign a value to a modulator directly because it will only exists the second time. Just test if one (any) modulator is not nil to assign the other modulators based on statedata.

**Reply by damiensellier** [reply]
I totally agree with Dominique, I never stored midi i/o settings as statedata. It's useless. It's only for custom parameter values. For exemple I store the last program used, so that when the plugin starts it's requesting  to the machine the last used program automatically, or for exemple the last size of the plugin(small HD or 4K)....  Those settings not available by default in ctrlr.
I would recommand users to save the panel with their own midi i/o values set up (as .panel files) .  Those settings are stored in the panel file at the top of the file in the "panel" node.
If you want to save those i/o settings within a vst host, I suggest selecting the midi i/o then save the plugin settings as "save as default preset" and everytime you launch the plugin it's going to keep the midi i/o settings from the default configuration.

As an exported stanndalone app, it's going to be stored in the  .settings file. so don't worry you have the midi setup stored anyway.

**Reply by Godlike-Productions** [reply]
Finally got it to work.  All the MIDI settings are strings, and you are correct with getProperty v getPropertyString.

getPropertyString does get something, but it's userdata.

I've updated my code above, and on the wiki.

The MIDI stuff doesn't seem to get saved when you just do a saveState.  I may have messed something up, but I needed to specifically write each one to the value tree.

**Comment by dobo365** [comment]
Thx Damien, I will look at those.
This is weird. I tried with an empty panel and put the statement in the console to get the error.
I also tried in a simple function and when running the function from the console I also get the error.

NOW: I just re-opened my simple empty panel and retried and it works!
Must have been some issue coming from the fact that I tried different syntaxes before finding the right way to assign the ValueTree.

FYI, you don't need to have properties defined; the simple initial assignment statement is working alone as well.

**Comment by dobo365** [comment]
OK. Now I can build and write a ValueTree with different levels and childs. I'm using toXmlString() to write the ValueTree in a text file.
When trying to read the ValueTree from the file, I'm trying to use fromXml() but it doesn't work.
Doing what(vtTest) in Ctrlr where vtTest is my ValueTree, I don't see fromXml() listed as method in 5.3.201

```
>>> 	vtTest = ValueTree("test")
what(vtTest)

Object type [ValueTree]
-----------------------------------------------------------------

Members:
	                    getSibling:	function
	                        __init:	function
	                       indexOf:	function
	              getChildWithName:	function
	                   toXmlString:	function
	                   getProperty:	function
	           removeAllProperties:	function
	                     moveChild:	function
	                      getChild:	function
	                getNumChildren:	function
	                   addListener:	function
	             removeAllChildren:	function
	                       hasType:	function
	      getOrCreateChildWithName:	function
	                    createCopy:	function
	                   hasProperty:	function
	                   setProperty:	function
	                       isValid:	function
	                   removeChild:	function
	                      addChild:	function
	                     createXml:	function
	                     getParent:	function
	                isEquivalentTo:	function
	                removeProperty:	function
	          getChildWithProperty:	function
	     sendPropertyChangeMessage:	function
	                 writeToStream:	function
	                       getType:	function
	                    isAChildOf:	function
	                removeListener:	function
	              getNumProperties:	function
	               getPropertyName:	function
```

Could it be that this came later in Juce? Or was it not integrated from Juce into Ctrlr (don't know how this is working, isn't it just adding the Juce library to make the function available in Ctrlr?)
**Any advice on how to read the ValueTree from a file?** I'm sure it is possible as Roman is reading the stateData that are ValueTree objects.

**Reply by damiensellier** [reply]
From the source files, Roman first create the XML tree from the file, then uses fromXml() to read the tree. I don't think you can read from the file directly without storing the content as binary or a non-xml format, there's probably a missing bridge between calling the file and reading it. There's probably another method to create the stream properly.

So I think you should first export your file with **createXml** and save as a file, instead of toXmlString() which is more intended for debugging inside ctrlr.

**fromXml()**
This isn't designed to cope with random XML data - **it should only be fed XML that was created by the createXml() method.**

**Reply by dobo365** [reply]
Yes, you are right Damien and I also saw that (and I was going to change my code if that was needed)
BUT
**the issue is that, as you can see above, createXML() is listed in the ValueTree functions but not fromXml()...**

Strange, isn't it?

**Reply by damiensellier** [reply]
yes it's strange. 
https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Lua/JuceClasses/LCore.cpp
line 855 to 893

//.def("createXml", &ValueTree::createXml)  is disabled and other classes are missing.

**Reply by damiensellier** [reply]
@dobo365 Dominique can you please post a sample panel with some of your valuetree code (save as file/load xml) so that I can experiment with it?

**Reply by dobo365** [reply]
Even stranger then as createXml() is well listed as available (when using what() ) while it seems commented in your extract above!
And indeed, no trace of fromXml()...

**Comment by dobo365** [comment]
Example:

```
function Hello()

	vtTest = ValueTree("Test")
	vtSub1 = ValueTree("Sub1")
	vtSub1:setProperty ("ID", 123, nil)
	vtSub2a = ValueTree("Sub2a")
	vtSub2a:setProperty ("value", "abc", nil)
	vtSub2b = ValueTree("Sub2b")
	vtSub2b:setProperty ("value", "def", nil)

	-- Assemble tree
	vtTest:addChild (vtSub1, -1, nil)
	vtSub1:addChild (vtSub2a, -1, nil)
	vtSub1:addChild (vtSub2b, -1, nil)

	-- Display tree
	console(vtTest:toXmlString())

	-- Export tree
	f = utils.saveFileWindow("Save", File(""), "*.txt", true)
	if f:isValid() then
		if f:existsAsFile()==false then
			f:create()
		end	
		f:replaceWithText(vtTest:toXmlString(), false, false)
	end

end
```

**Reply by damiensellier** [reply]
@dobo365 Salut Dominique. I tried with 5.6.23, which has the ValueTree Juce Class implemented the way described above.
It doesn't work, toXmlString() gives an error : 

```
Name: toXmlString
Error message: No matching overload found, candidates:
String toXmlString(ValueTree const&,custom [N4juce10XmlElement10TextFormatE] const&)

```

I tried with 5.3.198 and it worked very well :) I'll compare the files for both ctrlr version to see how I can get the best of both world.
I'm pretty sure I can add to the Ctrlr source a fromXmlString function to load a file as string.
They talk about it on Juce forum:
https://forum.juce.com/t/valuetree-y-u-no-exist/33464

I will also enable the createXml for luaBind and the other missing function from Juce if it's possible. It's in my to do list for my fork. I'll let you know if when I compile, it's working. ValueTree is a very powerful tool. it can easily replace properties on the fly and can produce XML which is a very convenient medium to store datas. It's perfect for preset Librarians.
This 1 step XML file to ValueTree function would be very neat.

**Reply by dobo365** [reply]
FYI, as this is not working, I'm now investigating XmlElement instead which seems to be fully implemented in 5.3.201.
Didn't succeed to have an output yet; Struggling with the syntax, as usual ;-)

BTW, Damien, would it be possible to compile and build 5.3.201 from the code but with VS2019 or Xcode?
Why would we need to still use VS2010? Due to the Juce version?
I don't like the new layout (colors) of the last version and would like to stay with the 5.3.201 version look.
I want also to have both VST2 and VST3.
And I read in another thread that all the memory block handling changed with this new version (due to Juce?) so it would need a lot of rework for me.
In summary: is it possible to get a working 5.3.201 or close to it so I could start from there and make my own version (just adding a few things but for sure keeping all the good that is available in it)?
Thx in advance for your insight about this.

**Reply by damiensellier** [reply]
I tried building 5.3.198 to just add the extra linker flag for pace ilok but I failed at it. So I had no choice but to go to the latest version and update my panels and so on. 
When trying to build the older version I had errors in Xcode, even with the corresponding Juce version from that time v4. I bet it was coming from non matching file path. So my bet is that it's totally possible to build it but requires a little bit of setup in the introjucer.
I don't know how Roman Kubiak handled it, some .xcodeproj are missing in the build folder (no Mac VST for exemple).
Even that old Juce version can create VST3 so it would be nice to have it.

For exemple the solid 5.3.198 is referenced as ctrlr 5.3.198 [8b32662] in the repository and requires JUCE 4.0.2. 
It requires starting from the .jucer file located in the build folder. 
This .jucer file doesn't work with projucer but with its predecessor introjucer.app located in the Juce 4.0.2 folder extra/introjucer/..

Roman exported VS2012 and VS2013 vcxproj files. 
Can you please post the "about" screencapture of your beloved Ctrlr 5.3.201 ? And we could start from there to go to the corresponding commit of the release day.

But just to let you know, you don't have to worry about the newer version vs the old ones. 
Appart from the colors I don't see any major differences. the 5.6.23 is solid and I use it nowadays.
The only one I could see is about sendMidiMessageNow instruction lines that requires adding MemoryBlock({}) like this:
panel:sendMidiMessageNow(CtrlrMidiMessage(MemoryBlock({0xf0, 0x0f, 0x40, 0x00, 0x00, 0x01, 0x00, 0x02, 0x00, v, 0xf7})))

Well and probably those ValueTree disrepancies we were talking about before.

And it's retrocompatible so you can develop your panel on the old Ctrlr and export with the new one without any issue.

**Reply by damiensellier** [reply]
I think the main problem is that luabind is not properly done for all the valueTree class.
according to JUCE:
toXmlString (const XmlElement::TextFormat &format={}) const

If we take the JuceClasses/LCore.cpp [Line 887](https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Lua/JuceClasses/LCore.cpp#L887)

`.def("toXmlString", &ValueTree::toXmlString)  `

It is wrong, it requires to pass the other variables (output format, header etc..) required by the JUCE function , hence the error.

EDIT: I'm starting to understand how LUA to JUCE works with the help of LuaBind, I think I'm going to start adding some function not available in Ctrlr via LuaBind.
Most of the function we use in LUA scripts are in [CtrlrLuaManager.cpp from line 859](https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Lua/CtrlrLuaManager.cpp#L859) with the namespace luabind. it won't be hard to mimic the ones available to add new Juce features we are missing on the terrain.

**Reply by Godlike-Productions** [reply]
You may need to add a constant at the end (try "\n"), like you do in the file export methods with 5.6.23 and later.  I think it just inserts this delimiter at the end of the line.

**Comment by Godlike-Productions** [comment]
The only downside of 5.6.23 is that it doesn't catch errors, so is prone to crashing and crashing the host.  You'll need to run functions in protected space yourself to avoid this, but that's probably a good practice to get into anyway.

**Reply by damiensellier** [reply]
Hi @Godlike-Productions . What do you mean by Catch errors?
I use 5.6.23 and when I have LUA errors I always get the red popup with the errors and what is responsible for it. Same in the console. Is this what you're talking about?


![popup error](https://user-images.githubusercontent.com/55164385/149838707-ef48f950-c515-4218-b7ed-425df1df749e.png)

![console error](https://user-images.githubusercontent.com/55164385/149838732-05785e93-f97d-4f00-9711-4172c051c851.png)

**Comment by Godlike-Productions** [comment]
Lots of stuff still causes crashes.  For example try to read a variable with a nil value and put it to a text box, and it crashes the host.

eg

```
  -- author
        local authorTemp = (clipboardBank[tempListboxValue+1].author)
        clipboardAuthorLabel:getComponent():setPropertyString ("uiLabelText", authorTemp)
``` 

This will crash ctrlr and the host if clipboardBank[n].author is nil.

**Reply by damiensellier** [reply]
Oh okay I get it. I'll use the safe mode from now on following your instructions. thanks
